Fix #5388: Generate deterministic CRDs

Donnerbart opened the pull request:
## Description
<!--
Thank a lot for taking time to contribute to Fabric8 <3!

Please provide a description of what your PR does providing a link (if applicable) to the issue it fixes. It is
really helpful for people who would review your code.
-->

Fixes #5388

I'm not sure if you want the provided `Complex` custom resource example as is in your code base, since it has nested `ObjectMetaData`, `StatefulSetSpec` and `ServiceSpec` and is tested against a checked in YAML file. So on every change on those K8s resources, the checked in file needs to be adapted (for example, non-breaking changes like adding new fields).

## Type of change
<!---
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/main/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CHEATSHEET.md) accordingly
 - [x] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [x] I tested my code in Kubernetes
 - [x] I tested my code in OpenShift

<!--
Integration tests (https://github.com/fabric8io/kubernetes-client/tree/master/kubernetes-itests)
Please check integration tests and provide/improve tests if applicable.

Open your PR in Draft mode and verify all of the applicable Checklist items before marking your pull request as ready for review
-->

andreaTP commented on crd-generator/api/src/main/java/io/fabric8/crd/generator/DeterministicObjectMapper.java:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    ObjectMapper build() {
+      final ObjectMapper mapper = JsonMapper.builder(new YAMLFactory()
```
This mapper is configured slightly differently from before, more specifically:

`YAML_MAPPER.configure(SerializationFeature.WRITE_NULL_MAP_VALUES, false)` and `YAML_MAPPER.configure(SerializationFeature.WRITE_EMPTY_JSON_ARRAYS, false);` are missing.

Donnerbart replied:
Both are deprecated API, so I looked how to configure it with up to date API.

`YAML_MAPPER.configure(SerializationFeature.WRITE_NULL_MAP_VALUES, false)` is `.withConfigOverride(Map.class, configOverride -> configOverride.setInclude(construct(NON_NULL, NON_NULL)))`.

`YAML_MAPPER.configure(SerializationFeature.WRITE_EMPTY_JSON_ARRAYS, false);` is `.serializationInclusion(NON_EMPTY)`.

andreaTP replied:
Nice 👍 , thanks for the explanation!

andreaTP commented on crd-generator/api/src/main/java/io/fabric8/crd/generator/DeterministicObjectMapper.java:
```diff
@@ -0,0 +1,206 @@
+          .enable(YAMLGenerator.Feature.MINIMIZE_QUOTES)
+          .enable(YAMLGenerator.Feature.ALWAYS_QUOTE_NUMBERS_AS_STRINGS)
+          .disable(YAMLGenerator.Feature.WRITE_DOC_START_MARKER))
+          .configure(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY, true)
```
Have you attempted to only add those configurations: `MapperFeature.SORT_PROPERTIES_ALPHABETICALLY` and `SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS` to the old mapper without defining an entirely new and custom one?
What are the invariants that we are missing?

Donnerbart replied:
The problem is that a lot of relevant elements of a CRD are not maps, but sets or other unsorted collections. The custom code is only taking care of collections.

`MapperFeature.SORT_PROPERTIES_ALPHABETICALLY` and `SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS` is already configured in my code, but it won't affect structures like enums, the additional printer columns section etc.

Donnerbart replied:
With all tests of the `crd-generator` the following collections are serialized:
```
com.fasterxml.jackson.databind.node.TextNode
io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceColumnDefinition
io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceDefinitionVersion
io.fabric8.kubernetes.api.model.apiextensions.v1.CustomResourceColumnDefinition
io.fabric8.kubernetes.api.model.apiextensions.v1.JSONSchemaProps
```
I've added custom converters for all of them, including v1 and v1beta1 variants.

andreaTP reviewed (commented):
@Donnerbart thanks a lot for taking the time to go through this!

I have a couple of questions but let me know your thoughts!

andreaTP commented:
LGTM but I wish @metacosm can do a final review here.

andreaTP commented:
@shawkins input would be great too, at least to verify that those changes are not affecting somehow the KC operator.

Donnerbart commented:
I've changed the comparison for `JSONSchemaProps` to use `type` first, since `id` doesn't seem to be populated (but still fallback to the `id` if `type` is the same value).

I also added more JavaDoc to the `DeterministicObjectMapper` and its builder class.

Donnerbart commented:
The license check is failing for the checked in YAML file in the test resources.
Since the file is used for a byte-wise comparison, it cannot be changed.

I see that there are already some exclusions for test resources:
```
            <exclude>kubernetes-client/src/test/resources/**/*</exclude>
            <exclude>kubernetes-client/src/test/resources/mockito-extensions/*</exclude>
            <exclude>**/src/test/resources/ssl/*</exclude>
            <exclude>**/src/test/resources/io/fabric8/java/generator/approvals/*.*.approved.txt</exclude>
            <exclude>**/it/**/expected/**/*.expected</exclude>
```
I don't want to change `<exclude>kubernetes-client/src/test/resources/**/*</exclude>` to match all modules, so I've added `<exclude>**/src/test/resources/*.yml</exclude>`, which is hopefully narrow enough to not violate the license requirements.

Donnerbart commented:
@shawkins That's a great finding! I've double checked our complex example, since there were for sure more nondeterministic fields. Looks like these were all properties, that are now sorted via `.configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true)`.

So indeed, with the correctly sorted enums that quite hacky way of adding the collection serializer is not needed anymore.

I also cleaned up the complex example, by adding simplified, but stable versions of the K8s models being used. This still provides enough properties, that were unsorted before, but won't break on future changes on the K8s model classes.

andreaTP commented:
@Donnerbart the CI failure on Win seems to be relevant, probably the encoding of the file is causing a difference.
I would be completely ok if, instead of comparing byte-by-byte we compare with `String::equal` the full lines.

This approach might also play nicer with the License check, e.g. you can include the License header in the committed file and skip it during the comparison.

Donnerbart commented:
@andreaTP Yeah, when I saw Windows I had the same thought. Will look into this first thing tomorrow (it's already end of workday for me).

I guess you don't want any helper dependency like Apache Commons or so in the test scope, so I'll write that comparison by hand. Will also look into adding the license header then.

andreaTP commented:
Thanks for getting back, no rush!

> I guess you don't want any helper dependency like Apache Commons or so in the test scope, so I'll write that comparison by hand.

I (personally) have no issues with adding helper dependencies to the test scope ...

Donnerbart commented:
* Removed the license check exclusion for YAML files in test resources.
* Added the license header to the checked in YAML file.
* Changed the file comparison to `BufferedReader::readLine` (that should handle all types of newlines).
* Skipped leading comments and newlines for file comparison.
* Improved some JavaDoc on the Complex example classes.

shawkins commented on crd-generator/api/src/test/java/io/fabric8/crd/generator/ResourcesTest.java:
```diff
@@ -42,14 +42,14 @@ public void shouldSupportMultiplePrinterColumns() {
     r.decorate(dec2);
     assertEquals(2, r.getDecorators().size());
-    r.getDecorators().stream().forEach(d -> System.out.println(d));
+    r.getDecorators().forEach(System.out::println);
```
It should be fine to start removing a log of the sys outs in tests.

Donnerbart replied:
I removed the sys out, would be great if you could re-run the tests.

shawkins reviewed (approved):
LGTM, just one minor comment.

metacosm commented on crd-generator/api/src/main/java/io/fabric8/crd/generator/CRDGenerator.java:
```diff
@@ -105,7 +120,10 @@ Map<String, AbstractCustomResourceHandler> getHandlers() {
   }
 
-  public CRDGenerator customResourceClasses(Class<? extends CustomResource>... crClasses) {
+  // this is public API, so we cannot change the signature, so there is no way to prevent the possible heap pollution
```
Not sure I understand the purpose fo this comment…

Donnerbart replied:
IntelliJ showed a warning for this method regarding heap pollution ([this blog article explains it quite well](https://arbitrary-but-fixed.net/teaching/java/2018/02/01/java-generic-arrays-varargs.html)). It comes down to the combination of generics and varargs.

I assume this is public API and the method signature cannot be changed, so there is nothing we can do to prevent the possible heap pollution. To get rid of the warning we could either:
1. Make the method `final` and annotate with `@SafeVarargs`.
1. Add a `@SuppressWarnings("unchecked")` annotation.

I wasn't sure if 1. is an option, since it would also change the method signature and may break custom `CRDGenerator` implementations. So I chose 2. as solution.

metacosm replied:
Thanks! Wasn't aware of this issue.

manusa reviewed (approved):
LGTM, thx!
